clash-cores/src/Clash/Cores/Crc/Internal.hs

```diff
@@ -0,0 +1,737 @@
+    withInitCrc i = mkSoftwareCrcFromParams (newParams { _crcInitial = i }) dataWidth
+    runCrc initial dat = digest $ feed (withInitCrc initial) dat
+    onehots n = shiftL 1 . fromIntegral <$> indices n
+    postProcess x = v2bv . reverse <$> transpose x
+    f = postProcess $ bv2v . (\p -> runCrc p 0) <$> onehots (SNat @(CrcWidth crc))
+    g = postProcess $ bv2v . (\p -> runCrc 0 p) <$> onehots dataWidth
+
+-- | Vertically partition a matrix
+partitionMat
+  :: KnownNat q
+  => SNat p
+  -> Vec n (BitVector (p * q))
+  -> Vec n (Vec p (BitVector q))
+partitionMat SNat mat = fmap unpack mat
+
+-- | Flatten vertically partitioned matrix
+unPartitionMat
+  :: KnownNat q
+  => KnownNat p
+  => Vec n (Vec p (BitVector q))
+  -> Vec n (BitVector (p * q))
+unPartitionMat mat = fmap pack mat
+
+-- | Helper type that contains the flattened @F@ and @G@ matrices for each lane
+data CrcLaneParams (crcWidth :: Nat) (dataWidth :: Nat) (nLanes :: Nat) where
+  -- | The final lane
+  CrcLane1
+    :: Vec crcWidth (BitVector (crcWidth + dataWidth))
+    -- ^ Flattened @F|G@ matrix for this lane
+    -> CrcLaneParams crcWidth dataWidth 1
+  -- | Not the final lane
+  CrcLaneN
+    :: 2 <= lane
+    => SNat lane
+    -- ^ Which lane this is
+    -> Vec crcWidth (BitVector (crcWidth + lane * dataWidth))
+    -- ^ Flattened @F|G@ matrix for this lane
+    -> CrcLaneParams crcWidth dataWidth (lane - 1)
+    -> CrcLaneParams crcWidth dataWidth lane
+
+deriving instance (KnownNat crcWidth, KnownNat dataWidth, KnownNat nLanes)
+  => Lift (CrcLaneParams crcWidth dataWidth nLanes)
+
+deriving instance (KnownNat crcWidth, KnownNat dataWidth, KnownNat nLanes)
+  => Show (CrcLaneParams crcWidth dataWidth nLanes)
+
+-- | No domain in 'CrcLaneParams'
+type instance TryDomain t (CrcLaneParams crcWidth dataWidth nLanes) = 'NotFound
+
+-- | Contains all necessary parameters for the hardware CRC implementation
+data CrcHardwareParams (crcWidth :: Nat) (dataWidth :: Nat) (nLanes :: Nat)
+  = CrcHardwareParams
+      { _crcDataWidthHW :: SNat dataWidth
+      , _crcNlanes :: SNat nLanes
+      , _crcBaseParams :: CrcParams crcWidth
+      , _crcLaneParams :: CrcLaneParams crcWidth dataWidth nLanes
+      , _crcResidues :: Vec nLanes (BitVector crcWidth)
+      }
+  deriving (Show, Lift)
+
+-- | No domain in 'CrcHardwareParams'
+type instance TryDomain t (CrcHardwareParams crcWidth dataWidth nLanes) = 'NotFound
+
+-- | This class is used to indicate a CRC has a derived hardware implementation
+--
+-- @nLanes@ indicates the maximum number of @dataWidth@ words you want to
+-- process in a single cycle. For example the stream could be byte-oriented,
+-- but processing is done @n@-bytes at a time.
+--
+-- Use 'Clash.Cores.Crc.deriveHardwareCrc' to create an instance. No instances
+-- should be implemented by hand, because a proper instance requires
+-- compile-time evaluation.
+class
+  ( KnownCrc crc, KnownNat dataWidth, 1 <= dataWidth, KnownNat nLanes, 1 <= nLanes)
```

Comment:
The `KnownNat` superclass constraints are no longer needed now that we have `SNat` arguments.

Reply:
Or were they left intentionally? So that `HardwareCrc` implies `KnownNat dataWidth` and `KnownNat nLanes`.

Reply:
These constraints mean that whenever a function takes a `HardwareCrc` argument, it is already a given that `(KnownCrc crc, KnownNat dataWidth, 1 <= dataWidth, KnownNat nLanes, 1 <= nLanes)`. This is nice for the functions `crcEngine` and `crcValidator`, which would otherwise still require those constraints in their type signature (well, okay, admittedly, `crcEngine` doesn't need `1 <= nLanes`, but that's really not a problem).

Reply:
Oh wait, I could have just said "Yes" :-)

Reply:
Exactly, that's why I kept it in. Just to be clear. There are no longer any open points?

Reply:
Yeah, I´m good. You too @DigitalBrains1 ?